### PR TITLE
Added the Von Mises Fisher distribution and factory

### LINKED ISSUE #3049: Need for Von Mises-Fisher distribution implementation
### What is the idea?

For large scale reliability problems, the Von Mises Fisher distribution could be very useful,
 see [here](https://www.sciencedirect.com/science/article/pii/S0951832023001539) an example of application of this density in high-dimensional reliability problems.

This distribution is a generalization of [Von Mises distribution](https://openturns.github.io/openturns/latest/user_manual/_generated/openturns.VonMises.html?highlight=von+mises#openturns.VonMises.__init__) already present in the library.

https://en.wikipedia.org/wiki/Von_Mises%E2%80%93Fisher_distribution

### Why is this needed?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by regislebrun**
For now, only the computePDF/computeLogPDF and getRealization/getSample methods are available. The factory is based on moments estimators. It fixes #3049.

This distribution is supported by the unit sphere in arbitrary dimension, here is the PDF in dimension 3 (each point of the unit sphere is scaled according to the PDF value):

Figure_1

## Summary by CodeRabbit

* **New Features**
  * Experimental Von Mises–Fisher distribution and factory with Python bindings and a visualization helper for unit-sphere PDFs.
* **Improvements**
  * Special-function routines handle nu = 0/1 edge cases with optimized early-return behavior.
* **Documentation**
  * Docs/autosummaries, docstrings, examples and bibliography updated for the new distribution and factory.
* **Tests**
  * New Python tests validating distribution behavior, sampling, statistics and factory estimation.
* **Configuration**
  * Default solver precision and max-iteration settings added for factory estimation.

### CODE REVIEW

File: python/src/VonMisesFisherFactory_doc.i [hunk 1]
```diff
@@ -0,0 +1,62 @@
+%feature("docstring") OT::VonMisesFisherFactory
+R"RAW(VonMisesFisher factory.
+
+.. warning::
+    This class is experimental and likely to be modified in future releases.
+    To use it, import the ``openturns.experimental`` submodule.
+
+Notes
+-----
+The estimation is based on the method of moments.
```
Comment (jschueller) [inline python/src/VonMisesFisherFactory_doc.i:8]: the notes section is confined after methods so less visible ; its preferable to keep the maths in the main section, and use the Notes section for detais like the resource map keys

File: python/src/VonMisesFisherFactory_doc.i [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    \begin{eqnarray*}
+      \displaystyle\Hat{\mu}_n &=& \frac{\bar{X}}{\|\bar{X}\|}\\
+      \displaystyle\Hat{\kappa}_n &\mbox{s.t.}& \frac{I_{\nu}(\Hat{\kappa}_n)}{I_{\nu-1}(\Hat{\kappa}_n)} = \|\bar{X}\|
+    \end{eqnarray*}
+
+where :math:`\bar{X}=\frac{1}{n}\sum_{i=1}^n X_i` is the empirical mean, :math:`\nu=\frac{p}{2}`, and
+:math:`I_{\nu}` is the modified Bessel function of the first kind of order :math:`\nu`.
+
+The resolution of the equation defining :math:`\Hat{\kappa}_n` can be tuned through
+`VonMisesFisherFactory-AbsolutePrecision`, `VonMisesFisherFactory-RelativePrecision`,
```
Comment (jschueller) [inline python/src/VonMisesFisherFactory_doc.i]: prefer a bullet list

File: python/src/VonMisesFisherFactory_doc.i [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+`VonMisesFisherFactory-ResidualPrecision`, `VonMisesFisherFactory-MaximumIteration` in :class:`~openturns.ResourceMap`.
+
+If :math:`\|\bar{X}\|=0`, the sample is supposed to be uniformly distributed over the sphere and the default
+distribution :math:`\kappa=0` is returned.
+
+See also
+--------
+DistributionFactory, VonMisesFisher)RAW"
```
Comment (jschueller) [inline python/src/VonMisesFisherFactory_doc.i]: experimental.VMF

File: python/src/VonMisesFisher_doc.i [hunk 4]
```diff
@@ -0,0 +1,86 @@
+%feature("docstring") OT::VonMisesFisher
+R"RAW(VonMisesFisher distribution.
```
Comment (jschueller) [inline python/src/VonMisesFisher_doc.i]: maybe we can use the full name: von Mises–Fisher

File: python/src/VonMisesFisher_doc.i [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    Default value is 0.0.
+
+See Also
+--------
+VonMisesFisherFactory
+
+Notes
+-----
+   The distribution is of dimension :math:`p` but is supported by the unit sphere :math:`\mathbb{S}_p` of intrinsic dimension :math:`p-1`.
```
Comment (jschueller) [inline python/src/VonMisesFisher_doc.i]: no indent

File: python/src/VonMisesFisher_doc.i (hunk 4 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i]: fixed

File: python/src/VonMisesFisher_doc.i (hunk 5 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i]: Fixed

File: python/src/VonMisesFisherFactory_doc.i (hunk 3 above)
Comment (regislebrun) [reply python/src/VonMisesFisherFactory_doc.i]: Fixed

File: python/src/VonMisesFisherFactory_doc.i (hunk 1 above)
Comment (regislebrun) [reply python/src/VonMisesFisherFactory_doc.i:8]: fixed

**Comment by m-balesdent** [comment]
@sdubreui would you be available to review this PR when it will be ready?

File: python/src/VonMisesFisherFactory_doc.i (hunk 1 above)
Comment (regislebrun) [reply python/src/VonMisesFisherFactory_doc.i:8]: fixed

File: python/src/VonMisesFisherFactory_doc.i (hunk 2 above)
Comment (regislebrun) [reply python/src/VonMisesFisherFactory_doc.i]: Fixed

**Review by jschueller — APPROVED** [review]

File: python/test/t_VonMisesFisher_std.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+distribution.setParameter(distribution.getParameter())
+ott.assert_almost_equal(
+    distribution.getParameter(), [0.707107, 0.707107, 5.0], 1e-5, 0.0
+)
+
+# Set with kappa=0 (uniform on sphere)
+distUniform = otexp.VonMisesFisher(ot.Point([1.0, 0.0, 0.0]), 0.0)
+assert distUniform.getDimension() == 3
+ott.assert_almost_equal(distUniform.getMean(), [0.0, 0.0, 0.0])
+assert math.isnan(distUniform.computeEntropy())
```
Comment (jschueller) [inline python/test/t_VonMisesFisher_std.py]: seems when kappa=0 entropy simplifies to -logNormalizationFactor_ avoiding nan/inf

Comment (regislebrun) [reply python/test/t_VonMisesFisher_std.py]: Good catch, fixed.

**Comment by jschueller** [comment]
we'd need another reviewer to get this accepted

**Comment by regislebrun** [comment]
@m-balesdent @l-brevault Do you have a few minute left to review this PR (ok, maybe more than a few minutes)

**Comment by m-balesdent** [comment]
> @m-balesdent @l-brevault Do you have a few minute left to review this PR (ok, maybe more than a few minutes)

Maybe @sdubreui could also review this PR.

**Comment by jschueller** [comment]
@regislebrun there are a few conflicts with master

File: python/src/VonMisesFisher_doc.i [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    \qquad \mathbf{x}\in\mathbb{S}_{p-1}
+
+where :math:`\boldsymbol{\mu}\in\mathbb{S}_{p-1}` is the mean direction,
+:math:`\kappa\ge 0` is the concentration parameter,
+:math:`I_\nu` is the modified Bessel function of the first kind of order
+:math:`\nu` and the normalization constant is:
+
+.. math::
+
+    C(\kappa) = \frac{(2\pi)^{p/2}\, I_{p/2-1}(\kappa)}{\kappa^{p/2-1}}.
```
Comment (l-brevault) [inline python/src/VonMisesFisher_doc.i]: There is an issue in the definition of the normalization constant, according to the code and to Wikipedia and Scipy documentation, it should be the inverse :

    C(\kappa) = \frac{\kappa^{p/2-1}}{(2\pi)^{p/2}\, I_{p/2-1}(\kappa)}.

File: python/src/VonMisesFisher_doc.i [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+
+When :math:`\kappa = 0`, the distribution is uniform over the unit sphere.
+
+The **mean** is:
```
Comment (l-brevault) [inline python/src/VonMisesFisher_doc.i:29]: To help the user (and avoid potential misunderstanding), it could be interesting to explain that: 
The expected value of the Von Mises–Fisher distribution is not on the unit hypersphere, but instead has a norm less than one.

File: python/src/VonMisesFisher_doc.i [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+.. math::
+
+    \mathbb{E}[\mathbf{X}] = \frac{I_{p/2}(\kappa)}{I_{p/2-1}(\kappa)}\,
+    \boldsymbol{\mu}.
+
+The **entropy** is:
+
+.. math::
+
+    \mathcal{H} = \log C(\kappa) - \kappa\,\frac{I_{p/2}(\kappa)}{I_{p/2-1}(\kappa)}.
```
Comment (l-brevault) [inline python/src/VonMisesFisher_doc.i]: According to the code and the definition, it misses a minus in front of the first term: 

\mathcal{H} = - \log C(\kappa) - \kappa\,\frac{I_{p/2}(\kappa)}{I_{p/2-1}(\kappa)}.

File: python/src/VonMisesFisher_doc.i [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+Sampling is performed using the fast rejection-sampling algorithm of
+Pinzón and Jung [pinzon2023]_.
```
Comment (l-brevault) [inline python/src/VonMisesFisher_doc.i:44]: For the user, it could be interesting to add here a sentence to explain the difference with the von Mises distribution (also provided in OpenTURNS), the relation between von Mises and Von Mises-Fisher and to explain that in the case p=2, the distribution reduces to the von Mises distribution on the unit circle but explain the main differences in implementation with Von Mises. 
You can also provide a link to send to the page of the von Mises distribution

File: python/src/VonMisesFisher_doc.i [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
```
Comment (l-brevault) [inline python/src/VonMisesFisher_doc.i:45]: Maybe, it could be interesting to specify in the documentation for the user that due to the specificities of this distribution only the computePDF/computeLogPDF and getRealization/getSample methods are available.

**Review by l-brevault — APPROVED** [review]
Thank you for the introduction of this new distribution. 
Some suggested documentation corrections, if you agree with it.

File: python/src/VonMisesFisher_doc.i (hunk 7 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i]: Oups, fixed!

File: python/src/VonMisesFisher_doc.i (hunk 8 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i:29]: fixed

File: python/src/VonMisesFisher_doc.i (hunk 9 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i]: Oups, fixed

File: python/src/VonMisesFisher_doc.i (hunk 10 above)
Comment (regislebrun) [reply python/src/VonMisesFisher_doc.i:44]: Fixed. For $d=2$ the von Mises-Fisher distribution is supported by the unit circle while the von Mises and as such it doesn't have a density wrt the Lebesgue measure of $\mathbb{R}^2$, while the von Mises distribution is the distribution of the angle wrt the $x$-axis of the von Mises-Fisher distribution.

**Comment by regislebrun** [comment]
@jschueller I still have a lint issue. According to my local setup:
The W504 warnings are not reproducing with the current flake8 setup — newer versions of flake8 (3.9+) removed the W504 check (it conflicts with PEP 8's preference for breaking before operators). No changes needed; these warnings are already resolved.
What can I do here?

**Comment by jschueller** [comment]
I guess your version of flake8 is not matching the latest from pypi.
I never noticed it could make a difference
